chore: patch crates to use main branch of iroh dependencies

ramfox opened the pull request:
This PR updates the following dependencies to use their main branches:

- `iroh` from `https://github.com/n0-computer/iroh.git`
- `iroh-blobs` from `https://github.com/n0-computer/iroh-blobs.git`

rklaehn commented on src/protocol.rs:
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
-    let lp = unsigned_varint::encode::u64(data.len() as u64, &mut buffer);
-    writer.write_all(lp).await?;
+    let data_len = data.len() as u64;
+    writer.write_u64_le(data_len).await?;
```
not that it matters that much, but why le?

dignifiedquire replied:
mostly a personal preference

rklaehn replied:
bao also uses le, so I guess it is consistent. Also I don't really mind.

flub replied:
heh, i had the same comment.  In theory these days with most machines being LE there is less byte-swapping involved, which can be an argument for it (and I assumed it was, turns out it wasn't even that :smile: )

rklaehn commented on src/protocol.rs:
```diff
@@ -112,53 +114,41 @@ pub async fn read_lp<'a, R: AsyncRead + futures::io::AsyncRead + Unpin, T: Deser
-pub async fn read_lp_data<R: AsyncRead + futures::io::AsyncRead + Unpin>(
+pub async fn read_lp_data<R: AsyncRead + Unpin>(
     mut reader: R,
     buffer: &mut BytesMut,
```
So do we still need this annoying buffer at all?

The issue I have is that when handing the remaining reader to the AsyncSliceDecoder, I have to use chain to take these bytes into account.

But when I do that I have no way to get the reader out again, meaning that I can only read one item...

dignifiedquire replied:
There should be no additional bytes in here anymore, see the `.take()` it now only reads the acutal required bytes, not more

flub commented on src/provider.rs:
```diff
@@ -312,7 +311,8 @@ async fn send_blob<W: AsyncWrite + Unpin>(
             let file = tokio::fs::File::open(&path).await?;
             let mut reader = tokio::io::BufReader::new(file);
-            tokio::io::copy(&mut reader, &mut writer).await?;
+            tokio::io::copy_buf(&mut reader, &mut writer).await?;
```
lol, love the lack of docs on this...

dignifiedquire replied:
yeah, so by reading the code, the TLDR is, use it if your sources provide AsyncBufRead, as that allows you to save on the additional buffer that `copy` creates.

dignifiedquire commented on src/protocol.rs:
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
+        ensure!(r != 0, "no more data available");
+        read += r;
     }
+    ensure!(size == read, "expected to read {} but read {}", size, read);
```
fixed

dignifiedquire commented on src/protocol.rs:
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
-        }
+        let r = reader.read_buf(buffer).await?;
+        ensure!(r != 0, "no more data available");
+        read += r;
```
fixed

dignifiedquire commented on src/protocol.rs:
```diff
@@ -81,29 +81,31 @@ pub async fn write_lp<W: AsyncWrite + Unpin>(writer: &mut W, data: &[u8]) -> Res
-    let lp = unsigned_varint::encode::u64(data.len() as u64, &mut buffer);
-    writer.write_all(lp).await?;
+    let data_len = data.len() as u64;
+    writer.write_u64_le(data_len).await?;
```
🤣

Commit:
chore: add patch for `iroh` dependencies

This PR updates the following dependencies to use their main branches:

- `iroh` from `https://github.com/n0-computer/iroh.git`
- `iroh-blobs` from `https://github.com/n0-computer/iroh-blobs.git`

Commit:
add lockfile
